File: scripts/generate_icons.py
#!/usr/bin/env python3
"""
Script para redimensionar el ícono de Sivarcasas a todos los tamaños
necesarios para Android (densidades mipmap) y para PWA (public/icons/).

Uso:
  python scripts/generate_icons.py <ruta_al_icono_fuente.png>

Requisito: pip install Pillow
"""

import sys
import os
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("ERROR: Pillow no está instalado. Ejecuta: pip install Pillow")
    sys.exit(1)

# Carpeta raíz del proyecto (la carpeta que contiene este script está en scripts/)
PROJECT_ROOT = Path(__file__).parent.parent

# Tamaños necesarios para Android (carpeta mipmap → tamaño en px)
ANDROID_ICONS = {
    "mipmap-mdpi":    48,
    "mipmap-hdpi":    72,
    "mipmap-xhdpi":   96,
    "mipmap-xxhdpi":  144,
    "mipmap-xxxhdpi": 192,
}

ANDROID_RES_DIR = PROJECT_ROOT / "android" / "app" / "src" / "main" / "res"

# Tamaños para PWA (carpeta public/icons/)
PWA_ICONS = {
    "icon-192x192.png": 192,
    "icon-512x512.png": 512,
    "apple-touch-icon.png": 180,
}

PWA_ICONS_DIR = PROJECT_ROOT / "public" / "icons"


def resize_and_save(source: Image.Image, output_path: Path, size: int):
    """Redimensiona la imagen fuente al tamaño dado (cuadrado) y la guarda."""
    output_path.parent.mkdir(parents=True, exist_ok=True)
    resized = source.resize((size, size), Image.LANCZOS)
    resized.save(output_path, format="PNG", optimize=True)
    print(f"  ✓ {output_path.relative_to(PROJECT_ROOT)}  ({size}x{size}px)")


def main():
    if len(sys.argv) < 2:
        print("Uso: python scripts/generate_icons.py <ruta_al_icono_fuente.png>")
        print("\nEjemplo:")
        print("  python scripts/generate_icons.py sivarcasaslogo.png")
        sys.exit(1)

    source_path = Path(sys.argv[1])
    if not source_path.is_absolute():
        source_path = PROJECT_ROOT / source_path

    if not source_path.exists():
        print(f"ERROR: No se encontró el archivo fuente: {source_path}")
        sys.exit(1)

    print(f"\n📷 Cargando imagen fuente: {source_path.name}")
    source = Image.open(source_path).convert("RGBA")
    print(f"   Tamaño original: {source.width}x{source.height}px")

    # ── Android icons ──────────────────────────────────────────────
    print("\n📱 Generando íconos para Android...")
    for folder, size in ANDROID_ICONS.items():
        ic_launcher_path      = ANDROID_RES_DIR / folder / "ic_launcher.png"
        ic_launcher_round_path = ANDROID_RES_DIR / folder / "ic_launcher_round.png"
        ic_launcher_fg_path   = ANDROID_RES_DIR / folder / "ic_launcher_foreground.png"
        resize_and_save(source, ic_launcher_path, size)
        resize_and_save(source, ic_launcher_round_path, size)
        resize_and_save(source, ic_launcher_fg_path, size)

    # ── PWA icons ──────────────────────────────────────────────────
    print("\n🌐 Generando íconos PWA...")
    for filename, size in PWA_ICONS.items():
        pwa_path = PWA_ICONS_DIR / filename
        resize_and_save(source, pwa_path, size)

    print("\n✅ ¡Todos los íconos generados exitosamente!\n")
    print("Próximo paso: ejecuta 'npx cap sync android' para sincronizar con Android Studio.")


if __name__ == "__main__":
    main()

File: scripts/generate_icons.js
#!/usr/bin/env node
/**
 * generate_icons.js — Sivarcasas
 * 
 * Genera todos los íconos necesarios para Android (mipmap-*) y PWA (public/icons/)
 * a partir de un archivo de imagen fuente de alta resolución.
 * 
 * Usa `sharp` que ya es dependencia del proyecto (package.json).
 * 
 * Uso:
 *   node scripts/generate_icons.js <ruta_al_icono_fuente>
 *
 * Ejemplo:
 *   node scripts/generate_icons.js sivarcasaslogo.png
 *   node scripts/generate_icons.js public/logo.webp
 */

const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

const projectRoot = path.join(__dirname, '..');

// Densidades Android → tamaño en px del mipmap
const ANDROID_ICONS = {
  'mipmap-mdpi':    48,
  'mipmap-hdpi':    72,
  'mipmap-xhdpi':   96,
  'mipmap-xxhdpi':  144,
  'mipmap-xxxhdpi': 192,
};

const ANDROID_RES_DIR = path.join(
  projectRoot, 'android', 'app', 'src', 'main', 'res'
);

// Íconos PWA
const PWA_ICONS = {
  'icon-192x192.png': 192,
  'icon-512x512.png': 512,
  'apple-touch-icon.png': 180,
};

const PWA_ICONS_DIR = path.join(projectRoot, 'public', 'icons');

async function resizeAndSave(sourcePath, outputPath, size) {
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  await sharp(sourcePath)
    .resize(size, size, { fit: 'cover', position: 'centre' })
    .png({ compressionLevel: 9 })
    .toFile(outputPath);

  const rel = path.relative(projectRoot, outputPath);
  console.log(`  ✓ ${rel}  (${size}×${size}px)`);
}

async function main() {
  const sourceArg = process.argv[2];
  if (!sourceArg) {
    console.error('Uso: node scripts/generate_icons.js <ruta_al_icono_fuente>');
    console.error('\nEjemplos:');
    console.error('  node scripts/generate_icons.js sivarcasaslogo.png');
    console.error('  node scripts/generate_icons.js public/logo.webp');
    process.exit(1);
  }

  const sourcePath = path.isAbsolute(sourceArg)
    ? sourceArg
    : path.join(projectRoot, sourceArg);

  if (!fs.existsSync(sourcePath)) {
    console.error(`ERROR: Archivo no encontrado: ${sourcePath}`);
    process.exit(1);
  }

  const meta = await sharp(sourcePath).metadata();
  console.log(`\n📷 Fuente: ${path.basename(sourcePath)} (${meta.width}×${meta.height}px)\n`);

  // Android icons
  console.log('📱 Generando íconos Android...');
  for (const [folder, size] of Object.entries(ANDROID_ICONS)) {
    const resDir = path.join(ANDROID_RES_DIR, folder);
    await resizeAndSave(sourcePath, path.join(resDir, 'ic_launcher.png'), size);
    await resizeAndSave(sourcePath, path.join(resDir, 'ic_launcher_round.png'), size);
    await resizeAndSave(sourcePath, path.join(resDir, 'ic_launcher_foreground.png'), size);
  }

  // PWA icons
  console.log('\n🌐 Generando íconos PWA...');
  for (const [filename, size] of Object.entries(PWA_ICONS)) {
    await resizeAndSave(sourcePath, path.join(PWA_ICONS_DIR, filename), size);
  }

  console.log('\n✅ ¡Todos los íconos generados!');
  console.log('\nPróximo paso: npx cap sync android\n');
}

main().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});
